RFC-0001-torch-function-for-methods.md

```diff
@@ -0,0 +1,177 @@
+## Abstract
+This RFC describes changes necessary to allow `__torch_function__` to be used by methods of `torch.Tensor` in an attempt to make subclassing more accessible to the users of the class. This entails making an API for subclass views public, and a change in the signature of `__torch_function__`.
+
+## Motivation and Scope
+Quoting [[1]], [[2]] and [[3]], the potential goals of this proposal are:
+
+1. Support subclassing `torch.Tensor` in Python
+2. Preserve `Tensor` subclasses when calling `torch` functions on them
+3. Preserve `Tensor` subclasses when calling `numpy` functions on them
+4. Use the NumPy API with PyTorch tensors (i.e. NumPy API calls dispatch to `torch` functions)
+5. Use the PyTorch API with `torch.Tensor`-like objects that are _not_ `Tensor` subclasses
+6. Reuse NumPy ufunc implementations directly from PyTorch
+7. Allow operations on mixed array types, e.g. `tensor + ndarray`
+8. Preserve `Tensor` subclasses when calling `Tensor` methods.
+9. Propagating subclass instances correctly also with operators, using views/slices/indexing/etc.
+10. Preserve subclass attributes when using methods or views/slices/indexing.
+11. A way to insert code that operates on both functions and methods uniformly (so we can write a single function that overrides all operators).
+
+We propose to solve this problem with the following changes to PyTorch:
+
+1. Make methods and operators of `torch.Tensor` go through the `__torch_function__` machinery.
+2. Add a `types` argument to `__torch_function__`, to make it match NumPy's `__array_function__`.
+3. Make `torch.Tensor._make_subclass` public API.
+4. Make `torch.Tensor` gain a generic implementation of `__torch_function__`.
+
+## Usage and Impact
+Once this proposal is merged, users of subclasses of `torch.Tensor` will have a much more streamlined experience. Namely, the following code example will work as-is, without the need for any further modification:
+
+```python
+class SubTensor(torch.Tensor):
+    a = 1
+
+t = SubTensor([1])
+s = t.sum()
+isinstance(s, SubTensor)  # True
+s.a  # 1
+i = t[0]
+isinstance(i, SubTensor)  # True
+i.a  # 1
+
+s2 = t + torch.Tensor(1)
+isinstance(s2, SubTensor)  # True
+s2.a  # 1
+
+s3 = torch.Tensor(1) + t
+isinstance(s3, SubTensor)  # True
+s3.a  # 1
+```
+
+Additionally, it will provide subclass authors hooks to run whenever methods or operators are called, and to modify the result to their specific use-case, perform logging, or otherwise change the result or the action of the method. For example:
+
+```python
+import logging
+
+class LoggingTensor(torch.Tensor):
+    def __torch_function__(self, func, types, args, kwargs):
+        logging.info(f"func: {func!r}, args: {args!r}, kwargs: {kwargs!r}")
+        return super().__torch_function__(
+            func,
+            tuple(
+                t for t in types if not issubclass(t, LoggingTensor)
+            ),
+            args,
+            kwargs
+        )
+```
+
```

Comment:
A discussion of performance is still missing. It should state the requirements:
- no impact on `Tensor` performance
- no deterioration of current `__torch_function__` overhead
- small (sub-us) overhead on `Tensor` subclasses
And whether that's achievable.

Here would be a good place (it falls under "Impact").

Reply:
>     * no impact on `Tensor` performance

This seems undoable, from what I understand, since `Tensor` methods that don't take arguments don't go through the processing of `Tensor` arguments, and they'll have to. In addition, for those that _do_ take arguments, `self` will have to be added in the list of processed arguments, so there will be a minor hit. This is regardless of the approach we take (`__torch_finalize__` will also have these drawbacks).

Reply:
Hmm, that seems important to estimate first, in a bit more detail than I did in gh-22402, since it could be a showstopper.

> since `Tensor` methods that don't take arguments don't go through the processing of Tensor arguments, and they'll have to

would be good to know which these are

> In addition, for those that _do_ take arguments, `self` will have to be added in the list of processed arguments, so there will be a minor hit.

An estimate for this impact would be good. Should be straightforward, by comparing overhead on current `__torch_function__` for functions with different number of arguments. Maybe that info is already in one of @ngoldbaum's PRs.

Reply:
> would be good to know which these are

These would be methods like `sum`, `prod`, and others that don't take a `Tensor`-like other than `self`.

Reply:
> An estimate for this impact would be good. Should be straightforward, by comparing overhead on current `__torch_function__` for functions with different number of arguments. Maybe that info is already in one of @ngoldbaum's PRs.

The problem with this approach is it takes into account the protocol overhead along with the checking for `Tensor` operands. In any case, I took a look at https://github.com/pytorch/pytorch/pull/27064#issuecomment-554601189 (@ngoldbaum, please correct me if there are more recent benchmarks), but the benchmarks are too all over the place to come up with an accurate number (the issue @peterbell10 mentioned). I'd ballpark the upper bound of performance hit to be in the range 0.1 to 1 microsecond, for `Tensor`-only operands.

Reply:
That's not the latest PR you're referencing: https://github.com/pytorch/pytorch/pull/32194 was the final version with inlining dispatch rather than using the decorator; overhead is <100ns (see PR description).

Reply:
Well, that's Python, but it doesn't have comparisons on input argument size. In any case, the overhead should be sub-50 ns.

Reply:
Okay, sounds sensible for parsing one more parameter. If you add a sentence on that to the new `Performance` section as the estimate for parsing `self`, then I think that's enough for now.